src/v/cluster/archival/ntp_archiver_service.cc

```diff
@@ -802,6 +826,206 @@ ss::future<> ntp_archiver::upload_until_term_change() {
+    // as soon as we can, rather than potentially waiting for segment
+    // uploads).
+    {
+        auto units = co_await ss::get_units(_uploads_active, 1);
+        co_await maybe_upload_manifest(upload_loop_prologue_ctx_label);
+        co_await flush_manifest_clean_offset();
+    }
+
+    // This variable is set by the previous iteration of the upload loop.
+    // It's only used if the legacy mode is disabled.
+    std::optional<std::chrono::milliseconds> carried_over_backoff;
+
+    while (may_begin_uploads()) {
+        // Handle backoff before acquiring units from the semaphore. This
+        // will allow housekeeping to continue working.
+        if (carried_over_backoff.has_value()) {
+            co_await ss::sleep_abortable(carried_over_backoff.value(), _as);
+        }
+
+        // Hold sempahore units to enable other code to know that we are in
+        // the process of doing uploads + wait for us to drop out if they
+        // e.g. set _paused.
+        vassert(!_paused, "may_begin_uploads must ensure !_paused");
+        auto units = co_await ss::get_units(_uploads_active, 1);
+        vlog(
+          _rtclog.trace,
+          "upload_until_term_change: got units (current {}), paused={}",
+          _uploads_active.current(),
+          _paused);
+
+        // Bump up archival STM's state to make sure that it's not lagging
+        // behind too far. If the STM is lagging behind we will have to read a
+        // lot of data next time we upload something.
+        vassert(
+          _parent.archival_meta_stm(),
+          "Upload loop: archival metadata STM is not created for {} archiver",
+          _ntp.path());
+
+        if (_parent.ntp().tp.partition == 0 && _topic_manifest_dirty) {
+            co_await upload_topic_manifest();
+        }
+
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto is_synced = co_await _parent.archival_meta_stm()->sync(
+          sync_timeout);
+        if (!is_synced.has_value()) {
+            // This can happen on leadership changes, or on timeouts waiting
+            // for stm to catch up: in either case, we should re-check our
+            // loop condition: we will drop out if lost leadership, otherwise
+            // we will end up back here to try the sync again.
+            continue;
+        }
+
+        auto [target_size, min_size] = get_segment_size_limits();
+        upload_candidate_search_parameters search_params(
+          _ntp,
+          _start_term,
+          target_size,
+          min_size,
+          // TODO: use proper quota
+          std::nullopt,
+          std::nullopt,
+          false,
+          manifest_upload_required());
+
+        auto upload_candidate_list = co_await _ops->find_upload_candidates(
+          _rtcnode, search_params);
+
+        auto throttle_upload_loop = [this, &carried_over_backoff]() {
```

Comment:
nit: the name here makes it sound like this method performs a sleep. How about calling it something like `adjust_next_throttle` or somesuch?